akka-actor/src/main/scala/akka/pattern/AskSupport.scala

```diff
@@ -150,36 +180,54 @@ final class AskableActorRef(val actorRef: ActorRef) extends AnyVal {
    * INTERNAL API: for binary compatibility
    */
   protected def ask(message: Any, timeout: Timeout): Future[Any] =
-    internalAsk(message, timeout, ActorRef.noSender)
+    internalAsk(_ ⇒ message, timeout, ActorRef.noSender)
 
   def ask(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalAsk(_ ⇒ message, timeout, sender)
+
+  def ask(message: ActorRef ⇒ Any)(implicit timeout: Timeout): Future[Any] =
+    internalAsk(_ ⇒ message, timeout, ActorRef.noSender)
+
+  def ask(message: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef): Future[Any] =
     internalAsk(message, timeout, sender)
 
   /**
    * INTERNAL API: for binary compatibility
    */
   protected def ?(message: Any)(implicit timeout: Timeout): Future[Any] =
-    internalAsk(message, timeout, ActorRef.noSender)
+    internalAsk(_ ⇒ message, timeout, ActorRef.noSender)
 
   def ?(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalAsk(_ ⇒ message, timeout, sender)
+
+  def ?(message: ActorRef ⇒ Any)(implicit timeout: Timeout): Future[Any] =
+    internalAsk(message, timeout, ActorRef.noSender)
+
+  def ?(message: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef): Future[Any] =
     internalAsk(message, timeout, sender)
```

Comment:
Unfortunately it seems that this won't work. We can't provide a single signature `?(ActorRef => Any)(implicit Timeout, ActorRef)`, because multiple overloads with default arguments are not allowed.

On the other hand current implementation fails miserably with "ambiguous reference to overloaded definition", due to erasure, which means that we need to choose to leave either `?(ActorRef => Any)(implicit Timeout, ActorRef)` or `?(ActorRef => Any)(implicit Timeout)`.

First option would require to explicitly specify an implicit sender, adding some cruft to using ask pattern from non actor context (which is one of the primary use cases of the pattern):
```scala
implicit val sender = ActorRef.noSender
target ? (Request(_))
```

Second one will make is impossible to pass a reference to sender.